runtime/caches/tiered.ts

```diff
@@ -0,0 +1,141 @@
+      return Promise.resolve({
+        /** [MDN Reference](https://developer.mozilla.org/docs/Web/API/Cache/add) */
+        add: (_request: RequestInfo | URL): Promise<void> => {
+          throw new Error("Not Implemented");
+        },
+        /** [MDN Reference](https://developer.mozilla.org/docs/Web/API/Cache/addAll) */
+        addAll: (_requests: RequestInfo[]): Promise<void> => {
+          throw new Error("Not Implemented");
+        },
+        /** [MDN Reference](https://developer.mozilla.org/docs/Web/API/Cache/delete) */
+        delete: async (
+          request: RequestInfo | URL,
+          options?: CacheQueryOptions,
+        ): Promise<boolean> => {
+          let deleteResponse;
+          for (const caches of openedCaches) {
+            deleteResponse = await caches.delete(request, options);
+          }
+          return deleteResponse ?? false;
+        },
+        /** [MDN Reference](https://developer.mozilla.org/docs/Web/API/Cache/keys) */
+        keys: (
+          _request?: RequestInfo | URL,
+          _options?: CacheQueryOptions,
+        ): Promise<ReadonlyArray<Request>> => {
+          throw new Error("Not Implemented");
+        },
+        /** [MDN Reference](https://developer.mozilla.org/docs/Web/API/Cache/match) */
+        match: async (
+          request: RequestInfo | URL,
+          options?: CacheQueryOptions,
+        ): Promise<Response | undefined> => {
+          let matched;
+          const indexOfCachesToUpdate = [];
+          for (const [index, caches] of openedCaches.entries()) {
+            matched = await caches.match(request, options).catch(() => null);
+            if (!matched) {
+              indexOfCachesToUpdate.push(index);
+              continue;
+            }
+
+            const expires = matched.headers.get("expires");
+            const isStale = expires ? !inFuture(expires) : false;
+
+            if (isStale) {
+              indexOfCachesToUpdate.push(index);
+            } else {
+              // found a match, no need to check the rest
+              break;
+            }
+          }
+          if (matched) {
+            updateTieredCaches(indexOfCachesToUpdate, request, matched);
+            return matched;
+          } else {
+            return undefined;
+          }
+        },
+        /** [MDN Reference](https://developer.mozilla.org/docs/Web/API/Cache/matchAll) */
+        matchAll: (
+          _request?: RequestInfo | URL,
+          _options?: CacheQueryOptions,
+        ): Promise<ReadonlyArray<Response>> => {
+          throw new Error("Not Implemented");
+        },
+        /** [MDN Reference](https://developer.mozilla.org/docs/Web/API/Cache/put) */
+        put: async (
+          request: RequestInfo | URL,
+          response: Response,
+        ): Promise<void> => {
+          const putPromises = openedCaches.map((caches) =>
+            caches.put(request, response.clone())
```

Comment:
You have cloned inside `updateTieredCaches` method, so you're doing two body copies here, can you ensure that the body is copied only once per cache?

Reply:
It is only copied once per request, isn't it? The updateTieredCaches is called in a different context. The put inside the updateTieredCaches is not the same in the TieredCaches, it is the put from each cache inside the tiered cache.

Reply:
that makes sense, can you check if the underlying caches implementation are avoiding clone since you're cloning it here?